[caffe2] Refactor Predictor to support CUDA

harrysummer opened the pull request:
Update 2:
The pull request now removes the thread-safe implementation. Only predictor CUDA support is included. See [comment](https://github.com/pytorch/pytorch/pull/6298#issuecomment-380674538) below.

---

Update:
The commit now put the threadsafe implementation into `Predictor`, and also includes Predictor CUDA support.

---
Original:

The ONNX interface is not thread safe. All threading using the same ONNX backend would share the same Caffe2 workspace. When multiple evaluation are doing concurrently, they would write to the same blobs.

This pull request deals with the problem by adding an interface method `run_in_new_workspace` to Caffe2 ONNX backend. When evaluating an image, we create a new workspace to store the intermediate and final results, while sharing the network weights by inheriting those blobs in init_net from the default workspace.

Note that this might not work for RNN case. An ultimate solution might be some sort of copy-on-write mechanism.

harrysummer committed:
add an additional interface to caffe2 onnx backend to support multithreaded evaluation.

yinghai reviewed (commented):
I think we should tuck the change inside the predictor. Right now, with this change, current predictor is really useless.

harrysummer commented:
@yinghai Thanks for you comment. I am currently refactoring the predictor code to make it support both GPU and CUDA context. I will put the multithreading support in predictor after that. Please leave this PR open. And I will try to update it asap.

yinghai reviewed (changes requested):
Temporary block

harrysummer committed:
refactor Predictor to support CUDA and threadsafe

harrysummer committed:
update predictor test

yinghai commented:
cc @salexspb for predictor

harrysummer committed:
update binaries/predictor_verifier.cc

harrysummer committed:
move getBlobs declaration to header to fix CI error

harrysummer committed:
Fix CI errors:
1. Add default value for threadsafe argument in Predictor::{run,run_map}
2. Add missing argument for PredictorRegistry()->Create (nullptr workspace)

harrysummer committed:
release python interpreter lock in onnx caffe2 cpp backend

harrysummer committed:
fix CI error in ios_caffe_predictor

harrysummer committed:
change CPUContext to caffe2::CPUContext to avoid CI error in ios_caffe_predictor.h

harrysummer committed:
Change Predictor usages in ios_caffe.cc accordingly

yinghai commented on caffe2/python/onnx/backend.py:
```diff
@@ -782,7 +782,7 @@ def prepare(cls, model, device='CPU', **kwargs):
         # Build the C++ backend
         # TODO: build a predictor that supports GPU
```
Remove this? :)

yinghai commented on binaries/predictor_verifier.cc:
```diff
@@ -40,10 +40,11 @@ void run() {
   VLOG(1) << "Init net: " << ProtoDebugString(init_net);
   LOG(INFO) << "Predict net: " << ProtoDebugString(predict_net);
-  auto predictor = caffe2::make_unique<Predictor>(init_net, predict_net);
+  auto predictor = std::move(PredictorRegistry()->Create("CPU", init_net, predict_net, nullptr));
```
Do we need `std::move` here? I think the return value is already rvalue.

yinghai commented on caffe2/core/predictor.cc:
```diff
@@ -51,80 +33,121 @@ const ::google::protobuf::RepeatedPtrField<::std::string>& getBlobs(
+
+  // Since the tensor is a member of blob, we should not delete the pointer.
+  // By given a customized deleter, the pointer won't be deleted on shared_ptr deconstruction.
+  return std::shared_ptr<TensorCPU>(blob->template GetMutable<TensorCPU>(), [](TensorCPU *){});
```
This is hacky. Why do we need to change the interface to return `std::shared_ptr<TensorCPU>` in the first place? I feel that doing this defeats the purpose of `shared_ptr`.

harrysummer replied:
The reason of changing the interface from raw pointer to `shared_ptr` is to make it compatible to thread-safe evaluation case, in which we temporarily create a new `Workspace` and new `Blob`'s for calculation. Therefore, we should take responsibility to deallocate the memory allocation we have made for those temporary `Tensor`'s, and `shared_ptr` is the only clean solution. Then for non-thread-safe case, in order to make them share the same interface, we have to do these hacks.

yinghai commented on caffe2/core/predictor_gpu.cc:
```diff
@@ -0,0 +1,101 @@
+std::shared_ptr<TensorCPU> copyOutputTensor(
+    Workspace* ws,
+    const std::string& name,
+    CUDAContext *context);
```
Redundant declaration?

yinghai commented on caffe2/onnx/backend_rep.cc:
```diff
@@ -1,31 +1,41 @@
-    predictor_ = caffe2::make_unique<caffe2::Predictor>(init_net_, pred_net_);
+    switch (pred_net_.device_option().device_type()) {
+      case caffe2::DeviceType::CPU:
+        predictor_ = std::move(caffe2::PredictorRegistry()->Create("CPU", init_net_, pred_net_, nullptr));
```
same here, is this `std::move` necessary?

yinghai reviewed (commented):
Thanks for doing this! Overall it looks great. I have some question about usage of `std::shared_ptr` and some nits.

harrysummer committed:
fix CI error by mmodifying GLPredictor as a subclass of PredictorBase instead of Predictor

yinghai reviewed (approved):
Thanks! There are some nits but no blockers.

harrysummer committed:
1. Change `run()` and `run_map()` interface from using `OutputTensorVector&` to 'OutputTensorVector*'
2. Change `getNet`, `getBlobs`, `shareInputTensor`, `extractOutputTensor`, `copyInputTensor`, and `copyOutputTensor` to Google coding style.
3. Remove the inline TODO in backend.py
4. Remove `std::move` for unique_ptr -> unique_ptr assignment.
5. Remove the redundant declaration of `copyOutputTensor` in `predictor_gpu.cc`

harrysummer committed:
add missing & to fix CI error

harrysummer committed:
Another & yet to add.

harrysummer committed:
yet another & to add

harrysummer committed:
change GLPredictor constructor to intialize Predictor<CPUContext>

harrysummer commented:
It seems the testing error in `caffe2-py2-gcc5-ubuntu16.04-test` is not causes by my change. I will close this pull request and re-open it to trigger CI build and test again.

salexspb commented:
Hey guys, what is the use-case here? 
Actually the plan was to open-source predictor we have internally. this predictor is really mobile one, it doesn't work with MetaNetDef format we use for server side and doesn't do a good job for multi-threaded inference. 

If we invest into this, should we actually just put more time toward OSS of the server side predictor and cleaning up the GPU path?

salexspb reviewed (changes requested):
requesting changes  for a more detailed design discussion about predictors (see my previous comment)

salexspb commented:
to give you extra context: for server side inference we have folly::ThreadLocalPtr being used to make it thread safe. @jerryzh168  implemented a custom version for Caffe2 and is about to ship it. 

once that is in, we should be able to open-source the code. What is the urgency for this on your side? We can also share some preliminary code with you once we strip out folly dependency. So you could get your self familiar with the code and contribute.

harrysummer commented:
Thanks @salexspb for your clarifying of the official plan. To my understanding, this pull request is not completely conflicts with your plan. It consists of two features:
1. **Make `Predictor`, as its name indicates, independent to the underlying device.** Both CPU and CUDA should be support. If your plan is making predictor a mobile one (CPU only), why not making it a `Predictor<CPUContext>` and inherit the `PredictorBase`? I don't think we should limit the potential of `Predictor`.
2. **Make network prediction thread-safe by allowing predictor to create and run net in a temporary workspace.** This might have the similar user-case as your about-to-ship feature of thread-safety. It sounds to me that your `folly::ThreadLocalPtr` would be used to create multiple predictor instances for each thread. This is almost as heavy as running different Caffe2 instances in terms of memory consumption. This pull request deals with multi-thread evalution in one predictor. This would also be very helpful when comes to Python-side evaluation with multiple threads from a single python Predictor instance. The implementation should be parallel with your about-to-ship feature.

BTW, I think current `Predictor` already support `MetaNetDef` format in its constructor.

I will stay with my own implementation, which works fine for me, at the moment, and keep up to the official design.

Best Regards.

harrysummer commented:
When I was playing with my thread-safe implementation, I found it has a serious performance drop to create new workspace every time predicting a new image. I have switched to instantiate multiple predictors in one onnx backend instead. Therefore, I will only preserve predictor CUDA support part of the change. As for multi-threading, we are looking forward to the official ThreadLocalPtr solution.

harrysummer committed:
remove thread-safe implementation

salexspb commented:
@harrysummer, let me clarify a bit more here. Basically we already  have a PredictorBase class internally. And there is GPU predictor as well. I am trying to avoid code replication. 
Regarding MetaNetDef support - it is not really there, the idea behind MetaNetDef is that weights are read from a DB, not from init_net. Storing weights protobufs has 2GB limitation which doesn't work for all models.

How urgent is this matter for you? I wonder if you could fork Predictor in your own project for now? 
@jerryzh168 is working on the OSS version of Predictor and we can prioritize the GPU version open-source to get you unblocked. 

Basically after ThreadLocalPtr change lands (which needs may be a few more days) we would need to do some minor code clean up and ready to move internal predictors out.

harrysummer commented:
@salexspb Thanks for your clarification. It's nice to hear that Caffe2 is going to support GPU Predictor 
 and very soon. As for my side, we won't be blocked for now, since my Predictor pool implementation is working. It's perfectly OK if this pull request won't finally get merged, I just want to share my work with the open-source community in case anyone needs it, but I also respect and will migrate to the official implementation after that ships.

salexspb commented:
We are working on open-sourcing predictor library with a unified pytorch 1.0 design, let me close this one for now.
